File: tools/validate_wave3_assets.mjs
#!/usr/bin/env node
/** Vérifie catalogue, trace, manifeste et fichiers de la vague ImageGen 3. */

import { createHash } from 'node:crypto';
import { existsSync, readFileSync, statSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import {
  ALL_SPECS, BATTLE_BACKGROUNDS, CREATURE_DETAILS, DECOR_PROPS,
  MAP_ICONS, PUBLIC_SPECS, TERRAIN_TEXTURES, wave3Summary,
} from './wave3_asset_specs.mjs';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const readJson = (file) => JSON.parse(readFileSync(resolve(ROOT, file), 'utf8'));
const sha256 = (file) => createHash('sha256').update(readFileSync(file)).digest('hex');
const fail = (message) => { throw new Error(message); };
const expected = wave3Summary();

for (const [label, actual, wanted] of [
  ['terrain', TERRAIN_TEXTURES.length, 12], ['décor', DECOR_PROPS.length, 61],
  ['lieux', MAP_ICONS.length, 13], ['combats', BATTLE_BACKGROUNDS.length, 6],
  ['références', CREATURE_DETAILS.length, 7], ['public', PUBLIC_SPECS.length, 92],
  ['total', ALL_SPECS.length, 99],
]) if (actual !== wanted) fail(`${label}: ${actual}, attendu ${wanted}`);

if (new Set(ALL_SPECS.map((row) => row.key)).size !== 99) fail('clefs du catalogue dupliquées');
if (ALL_SPECS.some((row) => ['carte_citadelle', 'carte_chateau'].includes(row.key))) {
  fail('clef conditionnelle absente du code ajoutée au catalogue');
}

const trace = readJson('docs/reference/IMAGEGEN-WAVE3-TRACE.json');
const byKey = new Map(trace.entrees.map((row) => [row.clef, row]));
if (trace.entrees.length !== 99 || byKey.size !== 99) fail('trace incomplète ou dupliquée');
const manifest = readJson('apps/client/public/img/manifeste.json');
const manifestByKey = new Map(manifest.entrees.map((row) => [row.clef, row]));
if (manifest.entrees.length !== 197 || manifestByKey.size !== 197) fail('manifeste attendu à 197 clefs uniques');

for (const spec of ALL_SPECS) {
  const row = byKey.get(spec.key) ?? fail(`trace absente: ${spec.key}`);
  for (const [field, wanted] of [
    ['id', spec.id], ['fichier', spec.file], ['categorie', spec.category],
    ['famille', spec.family], ['largeur', spec.width], ['hauteur', spec.height],
    ['alpha', spec.alpha], ['repetable', spec.repeatable], ['invite', spec.prompt],
  ]) if (row[field] !== wanted) fail(`${spec.key}: ${field} diffère du catalogue`);
  if (spec.sourceReference && row.referenceSource !== spec.sourceReference) {
    fail(`${spec.key}: référence source différente`);
  }
  for (const id of [row.generationId, row.generationIdFinal, ...row.generationIdsExtractionAlpha, ...row.generationIdsRegeneration]) {
    if (!/^exec-[0-9a-f-]+$/.test(id ?? '')) fail(`${spec.key}: identifiant ImageGen invalide`);
  }
  const target = resolve(ROOT, spec.file.startsWith('docs/') ? spec.file : `apps/client/public/img/${spec.file}`);
  if (!existsSync(target)) fail(`${spec.key}: fichier absent`);
  if (row.octets !== statSync(target).size) fail(`${spec.key}: taille de trace périmée`);
  if (row.sha256 !== sha256(target)) fail(`${spec.key}: SHA-256 de trace périmé`);
  if (!spec.file.startsWith('docs/')) {
    const current = manifestByKey.get(spec.key) ?? fail(`manifeste absent: ${spec.key}`);
    for (const field of ['fichier', 'categorie', 'largeur', 'hauteur', 'octets', 'invite', 'generationId', 'sha256']) {
      if (current[field] !== row[field]) fail(`${spec.key}: divergence trace/manifeste sur ${field}`);
    }
    if (Boolean(current.repetable) !== spec.repeatable) fail(`${spec.key}: repetable invalide`);
  }
}

const manifestBytes = manifest.entrees.reduce((sum, row) => sum + row.octets, 0);
if (manifestBytes > manifest.budgetOctets) fail(`budget dépassé: ${manifestBytes}`);
console.log(JSON.stringify({
  ...expected,
  manifestEntries: manifest.entrees.length,
  manifestBytes,
  budgetBytes: manifest.budgetOctets,
  traceBytes: trace.entrees.reduce((sum, row) => sum + row.octets, 0),
  errors: [],
}, null, 2));

File: tools/validate_wave3_assets.py
#!/usr/bin/env python3
"""Valide formats, dimensions, alpha et raccords réels de la vague 3."""

from __future__ import annotations

import json
from pathlib import Path

from PIL import Image, ImageChops


ROOT = Path(__file__).resolve().parents[1]
PUBLIC = ROOT / "apps" / "client" / "public" / "img"


def seam_rmse(image: Image.Image) -> float:
    rgb = image.convert("RGB")
    diffs = [
        ImageChops.difference(rgb.crop((0, 0, 1, rgb.height)), rgb.crop((rgb.width - 1, 0, rgb.width, rgb.height))),
        ImageChops.difference(rgb.crop((0, 0, rgb.width, 1)), rgb.crop((0, rgb.height - 1, rgb.width, rgb.height))),
    ]
    sq = count = 0
    for diff in diffs:
        histogram = diff.histogram()
        for value in range(256):
            for channel in range(3):
                frequency = histogram[channel * 256 + value]
                sq += frequency * value * value
                count += frequency
    return (sq / max(1, count)) ** 0.5


def main() -> None:
    trace = json.loads(
        (ROOT / "docs" / "reference" / "IMAGEGEN-WAVE3-TRACE.json").read_text(encoding="utf-8")
    )["entrees"]
    if len(trace) != 99 or len({row["clef"] for row in trace}) != 99:
        raise ValueError("trace vague 3 incomplète ou dupliquée")
    errors: list[str] = []
    seams: dict[str, float] = {}
    for row in trace:
        path = ROOT / row["fichier"] if row["fichier"].startswith("docs/") else PUBLIC / row["fichier"]
        with Image.open(path) as image:
            if image.format != "WEBP":
                errors.append(f"{row['clef']}: format {image.format}")
            if image.size != (row["largeur"], row["hauteur"]):
                errors.append(f"{row['clef']}: dimensions {image.size}")
            if row["alpha"]:
                if image.mode != "RGBA":
                    errors.append(f"{row['clef']}: mode {image.mode}")
                else:
                    extrema = image.getchannel("A").getextrema()
                    if extrema[0] != 0 or extrema[1] < 250:
                        errors.append(f"{row['clef']}: alpha {extrema}")
            if row["repetable"]:
                value = seam_rmse(image)
                seams[row["clef"]] = round(value, 4)
                if value > 4:
                    errors.append(f"{row['clef']}: raccord RMSE {value:.4f}")
    print(json.dumps({"assets": len(trace), "seams": seams, "errors": errors}, ensure_ascii=False, indent=2))
    if errors:
        raise SystemExit(1)


if __name__ == "__main__":
    main()
